File: interview/LeetCode/EasyProblem/Python (1)/136_SingleNumber.py
# Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.

# Solution 1: Time complexity O(n^2) and Space complexity O(n)
# Algorithms: Iterate over all the elements in nums. If some number in nums is new to array, append it. Else If some number is already in the array, remove it
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        """
        :type nums: List[int]
        :rtype: int
        """
        no_duplicate_list = []
        for i in nums:
            if i not in no_duplicate_list:
                no_duplicate_list.append(i)
            else:
                no_duplicate_list.remove(i)
        return no_duplicate_list.pop()
    

File: interview/LeetCode/EasyProblem/C++/136_SingleNumber.cpp
// Copyright (c) hoduchieu01
/*
    Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.  
    Algorithm: Because of a xor a = 0 and a xor 0 = a, we can xor all numbers in array together to find the unique number.
    Time Complexity: O(n) and Space Complexity: O(n)
*/
int singleNumber(int* nums, int numsSize){
    int ans = 0;
    for (int i=0; i <= numsSize; i++){
        ans ^= nums[i];
    }
    return ans;
}
